Stratis.Bitcoin/Utilities/AsyncLock.cs

```diff
@@ -7,53 +7,105 @@
+            }
+
+            /// <inheritdoc />
+            public void Dispose()
+            {
+                this.toRelease.semaphore.Release();
+            }
+        }
+
+        /// <summary>Internal synchronization primitive used as a mutex to only allow one thread to occupy the critical section.</summary>
         private readonly SemaphoreSlim semaphore;
+
+        /// <summary>Helper object that allows implementation of disposable lock for convenient use with <c>using</c> statement.</summary>
+        /// <remarks>We use the disposable interfaced in a task here to avoid allocations on acquiring the lock when it is free.</remarks>
         private readonly Task<IDisposable> releaser;
 
+        /// <summary>
+        /// Initializes an instance of the object.
+        /// </summary>
         public AsyncLock()
         {
             this.semaphore = new SemaphoreSlim(1, 1);
             this.releaser = Task.FromResult((IDisposable)new Releaser(this));
         }
 
         /// <summary>
-        /// Usage: using(await myAsyncLock.LockAsync())
+        /// Acquires the lock.
         /// </summary>
+        /// <param name="cancel">Cancellation token that can be used to abort waiting for the lock.</param>
+        /// <returns>Disposable interface to enable using construct. Disposing it releases the lock.</returns>
         public Task<IDisposable> LockAsync(CancellationToken cancel = default(CancellationToken))
         {
             Task wait = this.semaphore.WaitAsync(cancel);
 
             if (wait.IsCompleted)
-            {
                 return this.releaser;
-            }
-            else
-            {
-                return wait.ContinueWith((_, state) =>
-                    (IDisposable)state,
-                    this.releaser.Result, cancel,
-                    TaskContinuationOptions.ExecuteSynchronously,
-                    TaskScheduler.Default);
-            }
+
+            return wait.ContinueWith((task, state) => (IDisposable)state, this.releaser.Result, 
+                cancel, TaskContinuationOptions.ExecuteSynchronously, TaskScheduler.Default);
         }
 
-        private sealed class Releaser : IDisposable
+        /// <summary>
+        /// Acquires the lock.
+        /// </summary>
+        /// <param name="cancel">Cancellation token that can be used to abort waiting for the lock.</param>
+        /// <returns>Disposable interface to enable using construct. Disposing it releases the lock.</returns>
+        public IDisposable Lock(CancellationToken cancel = default(CancellationToken))
         {
-            private readonly AsyncLock toRelease;
-
-            internal Releaser(AsyncLock toRelease)
-            {
-                this.toRelease = toRelease;
-            }
-
-            public void Dispose()
-            {
-                this.toRelease.semaphore.Release();
-            }
+            this.semaphore.Wait(cancel);
+            return this.releaser.Result;
```

Comment:
`this.releaser.GetAwaiter().GetResult()` to avoid aggregate exception wrapper.

Reply:
'.Wait(cancel)' will throw so why is this an issue?

Reply:
i don't think GetAwaiter().GetResult() is useful here, the releaser task is finished (it has always been)
i'll test it thoroughly to be sure